### ISSUE TITLE
No need to remove the void-class during training

### ISSUE THREAD

**Issue by gaopeng-eugene** [DrSleep/tensorflow-deeplab-resnet#31]
I think the following is trying to remove label 255.
    raw_prediction = tf.reshape(raw_output, [-1, n_classes])
    label_proc = prepare_label(label_batch, tf.pack(raw_output.get_shape()[1:3]), one_hot=False) # [batch_size, h, w]
    raw_gt = tf.reshape(label_proc, [-1,])
    indices = tf.squeeze(tf.where(tf.less_equal(raw_gt, n_classes - 1)), 1)
    gt = tf.cast(tf.gather(raw_gt, indices), tf.int32)
    prediction = tf.gather(raw_prediction, indices)

However, tf.one_hot can do this automatically.
   input_batch = tf.one_hot(input_batch, depth=n_classes)

The following code is for testing:
a = tf.placeholder(tf.uint8)
out = tf.one_hot(a,depth=21)
sess = tf.Session()
sess.run(out,feed_dict={a:7})
array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  1.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.], dtype=float32)
sess.run(out,feed_dict={a:67})
array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.], dtype=float32)

As we can see, the loss is 0 when label is larger than 21 which will not contribute to gradient under cross-entropy loss.

These two approaches seem to be the same but use one_hot encoding can achieve better performance in my experiment. About 3 point improvement for the vgg16 network.

**Comment by DrSleep** [comment]
If you use one-hot encoding, then I assume that you also use softmax_cross_entropy_with_logits as a loss function. If that is the case, then from the [documentation](https://www.tensorflow.org/versions/r0.12/api_docs/python/nn/classification#sparse_softmax_cross_entropy_with_logits):

> All that is required is that each row of labels is a valid probability distribution. If they are not, the computation of the gradient will be incorrect.

which is obviously not true, when your row contains only 0's.

On the other hand, when using sparse_softmax_cross_entropy_with_logits, one must explicitly ignore the void label since we do not want to penalise predictions at the corresponding pixels.

**Comment by gaopeng-eugene** [comment]
you are right.
